proposals/xxxx-package-manager-manifest-api-redesign.md

```diff
@@ -0,0 +1,377 @@
+conventions-compliant API that we can rely on in the future.
+
+## Proposed solution
+
+This is how a regular manifest looks like currently:
+
+```swift
+let package = Package(
+    name: "Paper",
+    targets: [
+        // Helper tool.
+        Target(
+            name: "tool",
+            dependencies: [
+                "Paper",
+                "SwiftyJSON"
+            ]),
+        // The main library.
+        Target(
+            name: "Paper",
+            dependencies: [
+                .ByName(name: "Basic"),
+                .Target(name: "Utility"),
+                .Product(name: "CHTTPParser", package: nil),
+            ])
+    ],
+    products: [
+        .Executable(name: "tool", targets: ["tool"]),
+        .Libary(name: "Paper", type: .static, targets: ["Paper"]),
+        .Libary(name: "PaperDy", type: .dynamic, targets: ["Paper"]),
+    ],
+    dependencies: [
+        .Package(url: "http://github.com/SwiftyJSON", majorVersion: 1),
+        .Package(url: "../CHTTPParser", majorVersion: 1),
+        .Package(url: "http://some/other/lib", Version(1, 2, 3)),
+    ]
+    exclude: [
+        "Tests/Fixtures",
+    ]
+)
+```
+
+A system package manifest:
+
+```swift
+let package = Package(
+    name: "Clibgit",
+    pkgConfig: "libgit2",
+    providers: [
+        .Brew("libgit2"),
+        .Apt("libgit2-dev"),
+    ]
+)
+```
+
+---
+
+### Case conventions:
+
+> Names of types and protocols are UpperCamelCase. Everything else is lowerCamelCase.
+
+APIs incorrectly using upper camel cases should be fixed:
+
+* `Product`:
+
+```swift
+enum Product {
+    static func executable(name: String, targets: [String]) -> Product
+    static func library(name: String, type: LibraryProduct.LibraryType? = nil, targets: [String]) -> Product
+}
+```
+
+* `Target.Dependency`:
+
+```swift
+class Target {
+    enum Dependency {
+        case target(name: String)
+        case product(name: String, package: String?)
+        case byName(name: String)
+    }
+}
+```
+
+* `Package.Dependency`:
+
+```swift
+final class Package {
+    class Dependency {
+        static func package(...) -> Dependency  // Represents all static methods having name `Package`.
+    }
+}
+```
+
+### `Product`:
+
+// I propose we turn the tagets parameter on the static methods variadic, so brackets can be ommited.  
+// This will also automatically disallow specifying empty arrays in targets.
+
+```swift
+enum Product {
+    static func executable(name: String, targets: String...) -> Product
+    static func library(name: String, type: LibraryProduct.LibraryType? = nil, targets: String...) -> Product
+}
+```
+
+call site:
+
+```swift
+products: [
+    .executable(name: "tool", targets: "tool"),
+    .libary(name: "Paper", type: .static, targets: "Paper", "Basic"),
+]
+```
+
+### `SystemPackageProvider`:
+
+// This is currently an enum with case name being system package manager carrying package name in their payload.  
```

Comment:
... *the* case name ... being *the* ... *which carries* ... *the* package name ... in its payload, or something like that